crml/sylo/src/migration.rs

```diff
@@ -0,0 +1,226 @@
+/* Copyright 2020 Centrality Investments Limited
+*
+* Licensed under the LGPL, Version 3.0 (the "License");
+* you may not use this file except in compliance with the License.
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+* You may obtain a copy of the License at the root of this project source code,
+* or at:
+*     https://centrality.ai/licenses/gplv3.txt
+*     https://centrality.ai/licenses/lgplv3.txt
+*/
+
+use crate::{device, groups, inbox, vault};
+use frame_support::{decl_error, decl_module, decl_storage, dispatch::Vec, ensure, weights::SimpleDispatchInfo};
+use frame_system::{ensure_root, ensure_signed};
+use sp_runtime::{DispatchError::BadOrigin, DispatchResult};
+
+pub trait Trait: device::Trait + groups::Trait + inbox::Trait + vault::Trait {}
+
+type DeviceId = device::DeviceId;
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		InputError,
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as SyloMigration {
+		MigrationAccount: T::AccountId;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin, system = frame_system {
+		type Error = Error<T>;
+
+		#[weight = SimpleDispatchInfo::FixedOperational(0)]
+		pub fn set_migrator_account(origin, account_id: T::AccountId) -> DispatchResult {
+			ensure_root(origin)?;
+			MigrationAccount::<T>::put(account_id);
+			Ok(())
+		}
+
+		#[weight = SimpleDispatchInfo::FixedOperational(0)]
+		pub fn remove_migrator_account(origin) -> DispatchResult {
+			Self::ensure_sylo_migrator(origin)?;
+			MigrationAccount::<T>::kill();
+			Ok(())
+		}
+
+		#[weight = SimpleDispatchInfo::FixedOperational(0)]
+		fn migrate_devices(origin, user_id: T::AccountId, device_ids: Vec<DeviceId>) -> DispatchResult {
+			Self::ensure_sylo_migrator(origin)?;
+			ensure!(device_ids.len() <= device::MAX_DEVICES, Error::<T>::InputError);
+
+			let mut devices = <device::Devices<T>>::get(user_id.clone());
+			ensure!(devices.len() + device_ids.len() <= device::MAX_DEVICES, Error::<T>::InputError);
+
+			for device_id in device_ids {
+				devices.push(device_id)
+			}
+
+			devices.sort();
```

Comment:
Worth checking with Sylo whether sorting here would affect the client, does seem unlikely...

Reply:
I think it's worth preserving the order. Would u accept this as an alternative:
```rust
for device_id in device_ids {
    if !devices.contains(&device_id) {
        devices.push(device_id);
    }
}
```